python_modules/libraries/dagster-aws/dagster_aws/pipes/clients/emr_serverless.py

```diff
@@ -0,0 +1,289 @@
+    def message_reader(self) -> PipesMessageReader:
+        return self._message_reader
+
+    @classmethod
+    def _is_dagster_maintained(cls) -> bool:
+        return True
+
+    @public
+    def run(
+        self,
+        *,
+        context: OpExecutionContext,
+        start_job_run_params: "StartJobRunRequestRequestTypeDef",
+        extras: Optional[Dict[str, Any]] = None,
+    ) -> PipesClientCompletedInvocation:
+        f"""Run a workload on AWS {AWS_SERVICE_NAME}, enriched with the pipes protocol.
+
+            Args:
+                context (OpExecutionContext): The context of the currently executing Dagster op or asset.
+                params (dict): Parameters for the ``{BOTO3_START_METHOD}`` boto3 {AWS_SERVICE_NAME} client call.
+                    See `Boto3 API Documentation <https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/{AWS_SERVICE_NAME_LOWER}/client/{AWS_SERVICE_NAME_LOWER}.html#{AWS_SERVICE_NAME_LOWER_KEBAB}>`_
+                extras (Optional[Dict[str, Any]]): Additional information to pass to the Pipes session in the external process.
+
+            Returns:
+                PipesClientCompletedInvocation: Wrapper containing results reported by the external
+                process.
+            """
+        with open_pipes_session(
+            context=context,
+            message_reader=self.message_reader,
+            context_injector=self.context_injector,
+            extras=extras,
+        ) as session:
+            start_job_run_params = self._enrich_start_params(context, session, start_job_run_params)
+            start_response = self._start(context, start_job_run_params)
+            try:
+                completion_response = self._wait_for_completion(context, start_response)
+                context.log.info(f"[pipes] {self.AWS_SERVICE_NAME} workload is complete!")
+                self._read_messages(context, completion_response)
+                return PipesClientCompletedInvocation(session)
+
+            except DagsterExecutionInterruptedError:
+                if self.forward_termination:
+                    context.log.warning(
+                        f"[pipes] Dagster process interrupted! Will terminate external {self.AWS_SERVICE_NAME} workload."
+                    )
+                    self._terminate(context, start_response)
+                raise
+
+    def _enrich_start_params(
+        self,
+        context: OpExecutionContext,
+        session: PipesSession,
+        params: "StartJobRunRequestRequestTypeDef",
+    ) -> "StartJobRunRequestRequestTypeDef":
+        # inject Dagster tags
+        tags = params.get("tags", {})
+        tags.update(
+            {
+                "dagster/run_id": context.run_id,
+            }
+        )
+        params["tags"] = tags
+
+        # inject env variables via --conf spark.executorEnv.env.<key>=<value>
+
+        dagster_env_vars = {}
+
+        if isinstance(self.context_injector, PipesEnvContextInjector):
```

Comment:
This should not be conditional on `PipesEnvContextInjector`. Other context injector types might require bootstrap parameter as well.